lib/puppet/util/package/version/pip.rb

```
@@ -0,0 +1,230 @@
+          else
+            if ["c", "pre", "preview"].include?(letter)
+              letter = "rc"
+            else
+              if ["rev", "r"].include?(letter)
+                letter = "post"
+              end
+            end
+          end
+        end
+
+        return [letter, number.to_i]
+      end
+
+      if !letter && number
+        letter = "post"
+        return [letter, number.to_i]
+      end
+    end
+
+    def parse_local_version(local_version)
+      return local_version.split(/[\\._-]/).map{|part| part =~ /[0-9]+/ && part !~ /[a-zA-Z]+/ ? part.to_i : part.downcase} if local_version
+    end
+
+    def compose_key(epoch, release, pre, post, dev, local)
+      release_key = release.reverse
+      release_key.each_with_index do |element, index|
+        break unless release_key.at(index) == 0
+        release_key.delete_at(index) unless release_key.at(index + 1) != 0
+      end
+      release_key.reverse!
+
+      if !pre && !post && dev
+        pre_key = -Float::INFINITY
+      else
+        pre_key = pre ? pre : Float::INFINITY
+      end
+
+      post_key = post ? post : -Float::INFINITY
+
+      dev_key = dev ? dev : Float::INFINITY
+
+      if !local
+        local_key = -Float::INFINITY
+      else
+        local_key = local.map{|i| (i.is_a? Integer) ? [i, ""] : [-Float::INFINITY, i]}
+      end
+
+      return [epoch, release_key, pre_key, post_key, dev_key, local_key]
+    end
+
+    def compare(this, other)
+      if (this.is_a? Array) && (other.is_a? Array)
+        this  << -Float::INFINITY if this.length < other.length
+        other << -Float::INFINITY if this.length > other.length
+
+        this.each_with_index do |element, index|
+          return compare(element, other.at(index)) if element != other.at(index)
+        end
+      elsif (this.is_a? Array) && !(other.is_a? Array)
+        raise "Cannot compare #{this} (Array) with #{other} (#{other.class}). Only ±Float::INFINITY accepted." unless other.abs == Float::INFINITY
```

this will raise RuntimeError, i think would be better to raise a specific error for this.